### DISCUSSION TITLE
holepunch用法提议

### DISCUSSION THREAD

**Discussion by wulalawulala** [libp2p/go-libp2p#2817] [Q&A] [answered]
当使用holepunch的时候能否像使用/p2p-circuit/p2p/这样，让两个节点自动通过中继交换信息，通过中继自动打洞比如改成/direct-circuit/p2p/然后两者就能建立连接

**Comment by Stebalien** [comment, marked as answer]
Unless I'm misunderstanding your question, that's how it works.

1. Peer A connects to peer B over `/p2p-circuit`.
2. Peer B tries to directly connect back to peer A.
3. If that fails, peer B uses the [DCUtR](https://github.com/libp2p/specs/blob/master/relay/DCUtR.md) protocol, attempting to create a direct connection.

However note, this has only been implemented for QUIC (UDP) at the moment, as TCP-based hole-punching is significantly more complex.

**Reply by MarcoPolo** [reply]
We support TCP. But it generally doesn't work as well:
<img width="1093" alt="Screenshot 2024-06-05 at 7 30 24 PM" src="https://github.com/libp2p/go-libp2p/assets/594035/9c04a498-1d68-42bc-95bd-cc68133db2d3">

From @dennis-tra's talk at IPFS camp 2022
